File: Algorithms/Sorting/heapsort.cpp
#include<iostream>
using namespace std;

void swap(int *a,int*b)
{
    int t=*a;
    *a=*b;
    *b=t;
}

void max_heapify(int a[],int n,int i)
{
    int l = 2*i+1;
    int r = 2*i+2;
    int largest;

    if(l<n && a[l]>a[i])
        largest=l;
    else
        largest=i;

    if(r<n && a[r]>a[largest])
        largest=r;

    if(largest!=i)
    {
        swap(&a[i],&a[largest]);
        max_heapify(a,n,largest);
    }
}

void build_max_heap(int a[],int n)
{
    for(int i=n/2 -1;i>0;i--)
        max_heapify(a,n,i);
}

void heapsort(int a[], int n)
{
    build_max_heap(a,n);
    for(int i=n-1;i>0;i--)
    {
        swap(&a[0],&a[i]);
        max_heapify(a,i,0);
    }
}

int main()
{
    int n,arr[50];
    printf("Enter size of array: ");
    scanf("%d",&n);
    printf("Enter elements: ");
    for(int i=0;i<n;i++)
        scanf("%d",&arr[i]);

    heapsort(arr, n);

    printf("Sorted elements: ");
    for(int i=0;i<n;i++)
        printf("%d ",arr[i]);
}

File: Algorithms/Heapsort/heapsort.py
# Python program for heap sort
def heapify(arr, n, i):
    largest = i 
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and arr[i] < arr[l]:
        largest = l

    if r < n and arr[largest] < arr[r]:
        largest = r

    if largest != i:
        arr[i],arr[largest] = arr[largest],arr[i] # swap

        heapify(arr, n, largest)

#Function to sort a list
def heapSortAlgorithm(li):
    n = len(li)

    # Create a maxheap.
    for i in range(n // 2 - 1, -1, -1):
        heapify(li, n, i)

    #get the elements one by one and swap
    for i in range(n-1, 0, -1):
        li[i], li[0] = li[0], li[i]
        heapify(li, i, 0)

#main code
li = [ 12, 41, 2, 76, 19, 11, 98, 34]
heapSortAlgorithm(li)
print ("Array after HeapSort is: ")
for i in range(len(li)):
    print ("%d" %li[i])
